src/ray/common/function_descriptor.h

```diff
@@ -0,0 +1,181 @@
+
+namespace ray {
+/// See `common.proto` for definition of `FunctionDescriptor` oneof type.
+using FunctionDescriptorType = rpc::FunctionDescriptor::FunctionDescriptorCase;
+/// Wrap a protobuf message.
+class FunctionDescriptorInterface : public MessageWrapper<rpc::FunctionDescriptor> {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  FunctionDescriptorInterface(rpc::FunctionDescriptor message)
+      : MessageWrapper(std::move(message)) {}
+
+  ray::FunctionDescriptorType Type() const {
+    return message_->function_descriptor_case();
+  }
+
+  virtual size_t Hash() const = 0;
+
+  virtual std::string ToString() const = 0;
+
+  template <typename Subtype>
+  Subtype *As() {
+    return reinterpret_cast<Subtype *>(this);
+  }
+};
+
+class DriverFunctionDescriptor : public FunctionDescriptorInterface {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  explicit DriverFunctionDescriptor(rpc::FunctionDescriptor message)
+      : FunctionDescriptorInterface(std::move(message)) {
+    RAY_CHECK(message_->function_descriptor_case() ==
+              ray::FunctionDescriptorType::kDriverFunctionDescriptor);
+  }
+
+  virtual size_t Hash() const {
+    return std::hash<int>()(ray::FunctionDescriptorType::kDriverFunctionDescriptor);
+  }
+
+  virtual std::string ToString() const { return "{type=DriverFunctionDescriptor}"; }
+};
+
+class JavaFunctionDescriptor : public FunctionDescriptorInterface {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  explicit JavaFunctionDescriptor(rpc::FunctionDescriptor message)
+      : FunctionDescriptorInterface(std::move(message)) {
+    RAY_CHECK(message_->function_descriptor_case() ==
+              ray::FunctionDescriptorType::kJavaFunctionDescriptor);
+    typed_message_ = &(message_->java_function_descriptor());
+  }
+
+  virtual size_t Hash() const {
+    return std::hash<int>()(ray::FunctionDescriptorType::kJavaFunctionDescriptor) ^
+           std::hash<std::string>()(typed_message_->class_name()) ^
+           std::hash<std::string>()(typed_message_->function_name()) ^
+           std::hash<std::string>()(typed_message_->signature());
+  }
+
+  virtual std::string ToString() const {
+    return "{type=JavaFunctionDescriptor, class_name=" + typed_message_->class_name() +
+           ", function_name=" + typed_message_->function_name() +
+           ", signature=" + typed_message_->signature() + "}";
+  }
+
+  std::string ClassName() const { return typed_message_->class_name(); }
+
+  std::string FunctionName() const { return typed_message_->function_name(); }
+
+  std::string Signature() const { return typed_message_->signature(); }
```

Comment:
Java's `JavaFunctionDescriptor` class names this `typeDescriptor`. But I think `signature` is a better name. Can you change Java code to make it consistent?

Reply:
OK.